lib/nerve/reporter/zookeeper.rb

```diff
@@ -43,15 +43,18 @@ def start()
 
     def stop()
       log.info "nerve: removing zk node at #{@full_key}" if @full_key
-      report_down
-      @@zk_pool_lock.synchronize {
-        @@zk_pool_count[@path] -= 1
-        # Last thread to use the connection closes it
-        if @@zk_pool_count[@path] == 0
-          log.info "nerve: closing zk connection to #{@path}"
-          @zk.close!
-        end
-      }
+      begin
+        report_down
+      ensure
+        @@zk_pool_lock.synchronize {
+          @@zk_pool_count[@path] -= 1
+          # Last thread to use the connection closes it
+          if @@zk_pool_count[@path] == 0
+            log.info "nerve: closing zk connection to #{@path}"
+            @zk.close! rescue @@zk_pool.delete(@path)
```

Comment:
zk_pool_count has already been decremented, so the only remaining piece of bookeeping is to make sure that zk_pool no longer has the path in it (so that we don't "re-use" a dead connection)